### DISCUSSION TITLE
Interpret.plot_predictions with transformed variables

### DISCUSSION THREAD

**Discussion by SanBertero** [bambinos/bambi#821] [Q&A] [answered]
Hey there!

I had a question regarding the use of bmb.intrepret when the model includes transformations (scale(X), log(X) or both).

By default Bambi seems to plot the covariable back into its original scale/values, while the prediction in its transformed scale. Is there a way to control this either way? Say, keep Y(X) in its original scale, or plot it as a function of the transformed X? The latter is manageable by adding a new column in my dataframe, but the former isn't so direct.

More importantly, I had a case come up where I wanted to normalize log(x). scale(log(x)) seems to work for the inference model, but then bmb.interpret.plot_predictions fails/returns an error. Is it something related to my input or a limitation of the function?

Attached is a short Jupyter notebook with a quick example if it helps.
[ExampleInterpret.zip](https://github.com/user-attachments/files/16087129/ExampleInterpret.zip)

Thank you!

**Comment by GStechschulte** [comment, marked as answer]
Hey @SanBertero thanks for the question and notebook. 

> Is there a way to control this either way? Say, keep Y(X) in its original scale, or plot it as a function of the transformed X?

To keep `Y` on its original scale, you could pass an inverse transform function to the `transforms` parameter, e.g.

```python3
# model_5.formula = Formula('scale(Y) ~ scale(X)')

y_mean = np.mean(Y)
y_std = np.std(Y)

inv_transform = lambda x: x*y_std + y_mean

fig, ax = bmb.interpret.plot_predictions(
    model_5,results_5,
    'X',
    pps=True,
    prob=0.9, 
    transforms={"scale(Y)": inv_transform}
)
```
![image](https://github.com/bambinos/bambi/assets/63432018/d1ec4d8e-9953-4a83-b07f-16bcb329ff15)

> I had a case come up where I wanted to normalize log(x). scale(log(x)) seems to work for the inference model, but then bmb.interpret.plot_predictions fails/returns an error

What version of Bambi are you using? We pushed changes to address this bug, but I am not sure it is in the latest release.

**Reply by SanBertero** [reply]
Thank you for the reply and example! That's great and quite helpful!

I'm using version 0.13.0, which is the latest release I understand (from Oct. 2023). I did just try again but installing from the GitHub (it shows up as bambi-0.13.1.dev28+g6180e73. However, it then fails to import so I couldn't get it to work.

![Screenshot 2024-07-03 173803](https://github.com/bambinos/bambi/assets/28990427/723cbb65-0117-4775-91b5-34b0d5342761)

**Reply by GStechschulte** [reply]
Ahh yes. Looks like you are running into issue #820. I hope to have this issue resolved tomorrow.

**Reply by SanBertero** [reply]
Awesome, thank you!

**Reply by GStechschulte** [reply]
@SanBertero we now have a PR up #822 to resolve the import error. If you need to use the library right now you can install from the branch

```bash
pip install git+https://github.com/GStechschulte/bambi.git@import-error
```

**Reply by SanBertero** [reply]
Thank you for the update! I now have an error when building the models which I assume is slightly related to the same issue.

![Screenshot 2024-07-04 142144](https://github.com/bambinos/bambi/assets/28990427/2ac01dda-ab20-482a-9295-ef7ada50d693)

**Reply by GStechschulte** [reply]
Apologies about that. Can you uninstall bambi and reinstall from the branch linked above. I made a change that should resolve this error.

**Reply by SanBertero** [reply]
Thank you, I now got farther out into the example.

I notice that now the model builds a deterministic 'mu' compared to the latest release, which I'll figure out how it affects things/examples later...

However, and sorry to keep bothering, now bmb.interpret.plot_predictions fails. It seems that the variable `pps=True` has been changed to something else. I can make it work if I remove or set it to False to see the mean deviation. The code provides a warning saying it's now been moved to "response", but it doesn't seem to be a valid keyword argument.

![Screenshot 2024-07-04 145103](https://github.com/bambinos/bambi/assets/28990427/5c6bed16-e897-4b77-acb2-6e8531828330)

**Reply by GStechschulte** [reply]
>  I notice that now the model builds a deterministic 'mu' compared to the latest release

That's correct. This was introduced in #804. It shouldn't affect your results in a negative way. We wrap all the parameters of the likelihood into a deterministic node to make model graphs easier to read.

> It seems that the variable pps=True has been changed to something else. I can make it work if I remove or set it to False to see the mean deviation. The code provides a warning saying it's now been moved to "response", but it doesn't seem to be a valid keyword argument.

The changing of the parameter names is expected due to the commits in the merged PR linked above. This is weird because I am running your example on the `import-error` branch with `pps=True` and I don't get an error. Mmmm.

Perhaps it is an `xarray` version problem. In my environment I have

```bash
(bambi_main) ➜  ~ pip list | grep xarray
xarray              2024.6.0
xarray-einstats     0.7.0
```

**Reply by GStechschulte** [reply]
By the way, thanks for reaching out with questions and problems. The communication and feedback are much appreciated.

**Reply by SanBertero** [reply]
Hmm... I don't know. Here's the error I get when running predict on my own (I switch kind="pps" to "response" to avoid the warning, but everything else is the same), if it helps.

![Screenshot 2024-07-04 152712](https://github.com/bambinos/bambi/assets/28990427/3c10c42c-6d16-4fc5-873e-49111b1abe47)

Ahh, let me check that about xarray... I'm on 2023.7.0...

**Reply by SanBertero** [reply]
That's right! I installed a newer version of xarray from the github (2024.6.1.dev32+ga86c3ff4) and now everything worked out! Thank you so much for the help. Seems to be a dependency thing
